### ISSUE TITLE
Expose byte range(s) for attributes?

### ISSUE THREAD

**Issue by Jayonas** [RazrFalcon/roxmltree#113]
I'm evaluating XML libraries for use on a project, and one thing that I'm looking for is the ability to get the byte ranges where various XML items came from in the original string/document, for use in error reporting.  I found `Node.range()`, which is exactly what I'm looking for.  However, attributes aren't `Node`s so they don't benefit from this API.  I found `Attribute.position()`, but it only provides the first byte rather than a range.  Would you consider adding a `range()` method to `Attribute`?  Or does something like that already exist and I missed it?

Even better -- would you consider adding separate ranges for an attribute's name and value?  It would be nice to be able to pinpoint one or the other in different kinds of error messages.  E.g. an "invalid attribute" error could pinpoint the name, whereas an "invalid value" error could pinpoint the value.  I don't personally care very much whether the name's range includes its prefix (if any) or not, but I realize that's a thing that would need to be considered.

Another library that I've looked at is your `xmlparser`, which is far and away the best I've seen in terms of pinpointing the ranges of various items, not just separate prefix/name/value ranges for attributes, but even separate name/prefix for elements, separate full/contents-only ranges for comments and CDATA, and so forth.  Though as you point out in its documentation, that library is very low-level and therefore very limited in actual XML processing and validation capabilities.  I noticed that you recently went so far as to deprecate `xmlparser` in favor of simply merging it into this `roxmltree` library.  For anyone that has been using `xmlparser` and the extra range information it provides, they might find it easier to transition to this library if it also made that extra information available.  The main hang up for me is attribute ranges, but I would also applaud exposing subranges of various types of `Node`s as well.

**Comment by Jayonas** [comment]
Created PR #119 for adding `Attribute.range()`.  I also marked `Attribute.position()` as deprecated; let me know if you'd rather not do that, or have any other concerns.

I also have [a branch](https://github.com/Jayonas/roxmltree/tree/attr_subranges) for adding subranges for an attribute's qname and value behind a new feature flag, and would be happy for any feedback.

Let me know if I missed any protocol or such.  Despite being a relatively experienced dev, I'm pretty new to contributing on GitHub like this.  E.g. I couldn't figure out how to link the PR to this issue, I guess only this repo's maintainers can do that?

**Comment by RazrFalcon** [comment]
Thanks. Looks good to me.

As for qname/value ranges, you have an interesting trick to reduce memory usage. I like it.
`div_range` is a bit misleading name. I immediately thought about `<div/>` and not `=`. I would rather rename it, but I don't have good candidates for now.
Also, I think we could do some tricks and maybe use `Range<u8>` or `Range<u16>` at most. What chances that attribute name is larger than 255 or 65535 characters? `usize` would be an overkill.
I think we can settle for `Range<u16>` for now with some guard rails to make sure ranges larger than 65535 are clamped. And I will accept such PR.

>I couldn't figure out how to link the PR to this issue, I guess only this repo's maintainers can do that?

I don't think you can "link" it. You can mention an issue by id/number in a PR and GitHub will auto-reference it. That's about it.

**Comment by Jayonas** [comment]
Good point that `Range<usize>` is actually overkill for this.  I think we can get by with just two `u8`s, and once I started thinking of them as separate values instead of a single `Range` then better names for them became much more obvious: `qname_len` and `eq_len`.  I submitted PR #120 for it.  Let me know if you think either or both of those need to be `u16` instead (maxing out a `u8` for either of them already seems pretty extreme to me, but I suspect that you have more experience seeing wacky XML data).

Also -- will your automated build/test workflow on GitHub need to be manually updated to build/test with the new feature enabled?  If so, I'm guessing that I won't be able to do that for you as part of the PR?

Thanks for your time and feedback, and accepting these PRs.  I'm happy to be able to have this functionality in my app without needing my own fork.
